### PR TITLE
[RELAY][FRONTEND]Onnx to relay frontend

### PR DESCRIPTION
**Pull request by zhreshold**
ONNX -> relay frontend migration. 
This is co-authered by @siju-samuel with initial attempt https://github.com/dmlc/tvm/pull/2245

- Based on the fact that relay will not depend on NNVM, all nnvm onnx modules are stripped out and put into `relay/frontend`
- Put possible reusable code into `common` module.

### CODE REVIEW

**Comment by siju-samuel** [comment]
@zhreshold I had some patches on top of this change. you can apply this.
[0001-bug-fixes-and-more-testcases-added.patch.zip](https://github.com/dmlc/tvm/files/2691451/0001-bug-fixes-and-more-testcases-added.patch.zip)

**Comment by tqchen** [comment]
@srkreddy1238 @nishi-t @Huyuwei @hlu1  please help review this PR

File: python/tvm/relay/frontend/onnx.py [hunk 1]
```diff
@@ -0,0 +1,1078 @@
...
+    @classmethod
+    def _impl_v1(cls, inputs, attr, params):
+        # get number of channels
+        out = AttrCvt(op_name=dimension_picker('conv'),
+                      transforms={
+                          'kernel_shape': 'kernel_size',
+                          'dilations': ('dilation', (0, 0)),
+                          'pads': ('padding', (0, 0), revert_caffe2_pad),
+                          'group': ('groups', 1)},
+                      custom_check=dimension_constraint())(inputs, attr, params)
```
Comment (srkreddy1238) [inline python/tvm/relay/frontend/onnx.py]: I think we shouldn't pass 3 inputs (if bias available).

File: python/tvm/relay/frontend/onnx.py [hunk 2]
```diff
@@ -0,0 +1,1078 @@
...
+        out = AttrCvt(
+            op_name=dimension_picker('conv', '_transpose'),
+            transforms={
+                'kernel_shape': 'kernel_size',
+                'dilations': ('dilation', (0, 0)),
+                'pads': ('padding', (0, 0), revert_caffe2_pad)
+            },
+            disables=['output_shape'],
+            extras={'use_bias': len(inputs) == 3},
+            custom_check=dimension_constraint())(inputs, attr, params)
```
Comment (srkreddy1238) [inline python/tvm/relay/frontend/onnx.py]: Same as above.

File: python/tvm/relay/frontend/onnx.py [hunk 3]
```diff
@@ -0,0 +1,1078 @@
...
+            method = "NEAREST_NEIGHBOR"
+        elif mode == b'linear':
+            method = "BILINEAR"
+        else:
+            raise ValueError("Invalid ONNX upsample mode: {}".format(mode))
+        attr = {'scale':int(scales[-1]), 'method':method, 'layout':'NCHW'}
+        return AttrCvt('upsampling')(inputs, attr)
+
+
+class Shape(OnnxOpConverter):
```
Comment (srkreddy1238) [inline python/tvm/relay/frontend/onnx.py]: The shape op is a workaround in NNVM. We may need to revisit in relay.
Can be disabled here and handled outside this PR.

File: python/tvm/relay/frontend/onnx.py [hunk 4]
```diff
@@ -0,0 +1,1078 @@
...
+        'GlobalMaxPool': Renamer('global_max_pool2d'),
+        'BatchNormalization': BatchNorm.get_converter(opset),
+        # 'InstanceNormalization'
+        # 'LpNormalization'
+        'Dropout': AttrCvt('dropout', {'ratio': 'rate'}, ignores=['is_test']),
+        'Flatten': Renamer('flatten'),
+        'LRN': LRN.get_converter(opset),
+
+        # defs/reduction
+        #'ReduceMax': AttrCvt('max', transforms={'axes': 'axis'}),
```
Comment (nishi-t) [inline python/tvm/relay/frontend/onnx.py]: Did you forget to delete to forgot above comment line?

File: python/tvm/relay/frontend/onnx.py (hunk 1 above)
Comment (zhreshold) [reply python/tvm/relay/frontend/onnx.py]: will strip to first 2 inputs

File: python/tvm/relay/frontend/onnx.py (hunk 3 above)
Comment (zhreshold) [reply python/tvm/relay/frontend/onnx.py]: will disable for now

File: python/tvm/relay/frontend/onnx.py (hunk 4 above)
Comment (zhreshold) [reply python/tvm/relay/frontend/onnx.py]: yes, will clean the comments

**Comment by zhreshold** [comment]
@srkreddy1238 @nishi-t Can you guys help review again?

File: python/tvm/relay/frontend/onnx.py [hunk 5]
```diff
@@ -0,0 +1,1077 @@
...
+            if len(outputs) == 1:
+                return outputs
+            # TODO(zhreshold): support dropout mask?
+            outputs = outputs[:-1]
+        return outputs
+
+def from_onnx(model,
+              shape=None,
+              dtype="float32"):
+    """Convert from ONNX"s model into compatible relay Function.
```
Comment (jroesch) [inline python/tvm/relay/frontend/onnx.py]:
```suggestion
    """Convert a ONNX model into an equivalent Relay function.
```

File: python/tvm/relay/frontend/onnx.py [hunk 6]
```diff
@@ -0,0 +1,1077 @@
...
+    Onnx graph is a python protobuf object. The companion parameters will be handled automatically.
```
Comment (jroesch) [inline python/tvm/relay/frontend/onnx.py]:
```suggestion
    ONNX graphs are represented as a Python Protobuf object. The companion parameters will be handled automatically.
```

File: tutorials/relay/from_onnx.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
+"""
+Compile ONNX Models
+===================
+**Author**: `Joshua Z. Zhang <https://zhreshold.github.io/>`_
+
+This article is an introductory tutorial to deploy ONNX models with Relay.
+
+For us to begin with, onnx module is required to be installed.
```
Comment (jroesch) [inline tutorials/relay/from_onnx.py]:
```suggestion
For us to begin with, ONNX package must be installed.
```

File: tutorials/relay/from_onnx.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+target = 'llvm'
+
+input_name = '1'
+shape_dict = {input_name: x.shape}
+sym, params = relay.frontend.from_onnx(onnx_model, shape_dict)
+
+with relay.build_config(opt_level=1):
+    graph, lib, params = relay.build(sym, target, params=params)
+
+######################################################################
```
Comment (jroesch) [inline tutorials/relay/from_onnx.py]: Can we maybe just show how to use the `create_executor` interface in the tutorial as well? its much simpler then building a graph runtime by hand

Comment (zhreshold) [reply tutorials/relay/from_onnx.py]: @jroesch  Can you please advice how to elegantly evaluate executor created with an unordered dict of weights?

```python

exec = relay.build_module.create_executor('graph', mod, tvm.cpu(0), 'llvm')
# params is a dict of {'name', tvm.ndarray} 
tvm_output = exec.evaluate(mode)(x, *params.values()).asnumpy()   # wrong order
```
How do I feed the executor with correct weights without changing the params dict to odict?

Comment (jroesch) [reply tutorials/relay/from_onnx.py]: @zhreshold I realize the current API does not actually support keyword argument style for parameters. 

I'm going to update it with a PR right now, it should work like this:

```
exec = relay.build_module.create_executor('graph', mod, tvm.cpu(0), 'llvm')
tvm_output = exec.evaluate(mode)(x, **params).asnumpy()
```

Does that seem elegant enough?

Comment (zhreshold) [reply tutorials/relay/from_onnx.py]: Yes, that would perfectly solve my current problem

Comment (zhreshold) [reply tutorials/relay/from_onnx.py]: Will update once the PR get merged.

Comment (jroesch) [reply tutorials/relay/from_onnx.py]: PR is merged.

**Comment by icemelon9** [comment]
@srkreddy1238 @jroesch @nishi-t please help review this PR again

**Comment by Huyuwei** [comment]
@zhreshold Should we put the tests under https://github.com/dmlc/tvm/tree/master/tests/python/frontend, and add it to CI https://github.com/dmlc/tvm/blob/master/tests/scripts/task_python_frontend.sh/#L30-L34?

**Comment by zhreshold** [comment]
@Huyuwei Seems like there is `tests/python/frontend` and `tests/python/relay/frontend` both including mxnet and keras frontend already, can you clarify the relationship between them?

**Comment by Huyuwei** [comment]
@zhreshold tests/python/frontend is the right place to hold frontend tests, tests/python/frelay/frontend is duplicated and should be removed. In fact, tests/python/frelay/frontend is not running in the CI
See https://github.com/dmlc/tvm/pull/2316#issuecomment-450215747

**Comment by icemelon9** [comment]
@Huyuwei I'll remove the duplicate tests in a separate PR. 
@zhreshold Let's use tests/python/frontend for frontend tests.

**Comment by zhreshold** [comment]
ok, working on that

**Pull request merged**